src/Lean/Elab/PreDefinition/PartialFixpoint/Induction.lean

```diff
@@ -50,6 +50,23 @@ def CCPOProdProjs (n : Nat) (inst : Expr) : Array Expr := Id.run do
   return insts
 
 
+def desugarOrder (predType : Expr) (body : Expr) (reduceRhs : Bool := false) (fixpointType : Option Bool) : MetaM (Expr) := do
+  if fixpointType.isNone then
+    throwError "Trying to apply lattice induction to a non-lattice fixpoint. Please report this issue."
+  if body.isAppOfArity ``PartialOrder.rel 4 then
+    let lhsTypes ← forallTelescope predType (fun ts _ =>  ts.mapM inferType)
+    let names ← lhsTypes.mapM (fun _ => mkFreshUserName `x)
+    let bodyArgs := body.getAppArgs
+    withLocalDeclsDND (names.zip lhsTypes) fun exprs => do
+      let mut applied := if fixpointType.get! then (bodyArgs[2]!, bodyArgs[3]!) else (bodyArgs[3]!, bodyArgs[2]!)
+      for e in exprs do
+        applied := (mkApp applied.1 e, mkApp applied.2 e)
+      if reduceRhs then
+        applied := (applied.1, ←whnf applied.2)
+      mkForallFVars exprs (←mkArrow applied.1 applied.2)
```

Comment:
```suggestion
        applied := (applied.1, (← whnf applied.2))
      mkForallFVars exprs (←mkArrow applied.1 applied.2)
```
I had to get used to that as well, but Leo strongly suggests always using parentheses around `← foo` in monadic code, even if not strictly required by the term syntax. (And spaces after `←`)